qemu/index.js

```diff
@@ -56,45 +74,84 @@ app.post('/machines', upload.array(), function (req, res) {
   var id = uuid.v4();
   var name = machineDescription.name + '-qemu-' + id;
   var port = getPort();
-  var cmd = `qemu-system-x86_64 \
+  var drive = machineDescription.drive;
+  createImage(drivePath + drive, drivePath + uuid.v4())
+    .then((image) => {
+      var cmd = `qemu-system-x86_64 \
     -nodefaults \
+    -snapshot \
     -name "${name}" \
     -m ${machineDescription.memory} \
     -smp ${machineDescription.cpu} \
     -machine accel=kvm \
-    -drive file=${drive},format=qcow2 \
+    -drive file=${image.name},format=qcow2 \
     -vnc :${port.vnc} \
     -usb -device usb-tablet \
     -net nic,vlan=0,model=virtio \
     -net user,vlan=0,hostfwd=tcp::${port.plaza}-:9090,hostfwd=tcp::${port.rdp}-:3389 \
+    -monitor unix:/data/${id}.socket,server,nowait \
     -vga qxl \
   `;
 
-  exec(cmd, () => {});
-
-  return res.json({
-    id: id,
-    name: name,
-    plazaPort: port.plaza,
-    rdpPort: port.rdp,
-    vncPort: port.vnc,
-    status: 'booting'
-  });
-
+      exec(cmd, () => {});
+
+      return res.json({
+        id: id,
+        name: name,
+        plazaPort: port.plaza,
+        rdpPort: port.rdp,
+        vncPort: port.vnc,
+        status: 'booting'
+      });
+    });
 });
 
 app.delete('/machines/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}', function (req, res) {
 
   var machineId = req.path.substr(10);
-
-  exec('PID=$(ps aux | grep "' + machineId + ' " | grep qemu | awk \'{ print $2 }\' | tr "\\n" " ") ; kill ${PID}',
-    () => {});
-
-  return res.json({
-    id: machineId,
-    status: 'stopping'
+  let socket = net.connect({path: '/data/' + machineId + '.socket'}, () => {
+    socket.write('system_powerdown\n', () => {
+      return res.json({
+        id: machineId,
+        status: 'stopping'
+      });
+    });
   });
+});
 
+app.post('/images', upload.array(), function (req, res) {
+  var params = req.body;
+
```

Comment:
We should check if params are correct before trying to use them.

Chek if params.buildFrom and params.iaasId are  defined